bash: don't install loadables dir for static build

veehaitch opened the pull request:
The dynamic build ships a bunch of sample loadable builtins and installs them to `$out/lib/bash`. When building statically with musl, the target fails as it cannot produce any of the dynamically linked loadable builtins.

Although the failure is not fatal, it also introduces an impurity due to the heavy parallelization: _sometimes_ the files `loadables.h` and `Makefile.sample` end up in `$out/lib/bash` depending on how fast the compilation of the loadable builtins fails. Try the following to reproduce the issue:

```
nix build --rebuild --cores 48 nixpkgs#pkgsStatic.bash
```

This commit skips building and installing the loadable builtins altogether if building statically.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wentasah reviewed (approved):
This looks good to me. I can reproduce the reported non-reproducibility. It disappears with this change applied. All made changes make sense to me. The comments are clear.

SuperSandro2000 commented on pkgs/shells/bash/5.nix:
```diff
@@ -87,8 +91,15 @@ stdenv.mkDerivation rec {
+    "--enable-static-link"
   ];
 
+  # The configure script option for static linking is called `--enable-static-link`,
+  # not `--enable-static`. As the latter is a substring of the former, the stdenv
+  # setup assumes the configure script has a `--enable-static` option. We pass the
+  # `--enable-static-link` option as a `configureFlags` if we build statically.
```
```suggestion
  # The configure script option for static linking is called `--enable-static-link`, not `--enable-static`.
  # As the latter is a substring of the former, the stdenv setup assumes the configure script has a `--enable-static` option.
  # We pass the `--enable-static-link` option as a `configureFlags` if we build statically.
```
way easier to read

SuperSandro2000 commented on pkgs/shells/bash/static-no-install-loadables-dir.patch:
```diff
@@ -0,0 +1,14 @@
+ 		infodir=$(infodir) htmldir=$(htmldir) DESTDIR=$(DESTDIR) $@ )
+ 	-( cd $(DEFDIR) ; $(MAKE) $(MFLAGS) DESTDIR=$(DESTDIR) $@ )
+ 	-( cd $(PO_DIR) ; $(MAKE) $(MFLAGS) DESTDIR=$(DESTDIR) $@ )
++ifeq (,$(findstring -static,$(STATIC_LD)))
```
Can you try upstreaming this, too?

alyssais commented on pkgs/shells/bash/5.nix:
```diff
@@ -87,8 +91,14 @@ stdenv.mkDerivation rec {
     "bash_cv_termcap_lib=libncurses"
   ] ++ lib.optionals (stdenv.hostPlatform.libc == "musl") [
     "--disable-nls"
+    "--enable-static-link"
```
musl != static.  There's dynamically linked musl, and non-musl static builds.  This should check `isStatic`.

veehaitch replied:
Thanks! Now checks for `isStatic`!

veehaitch committed:
bash: don't install loadables dir for static build

The dynamic build ships a bunch of sample loadable builtins and installs
them to `$out/lib/bash`. When building statically with musl, the target
fails as it cannot produce any of the dynamically linked loadable
builtins.

Although the failure is not fatal, it also introduces an impurity due to
the heavy parallelization: _sometimes_ the files `loadables.h` and
`Makefile.sample` end up in `$out/lib/bash` depending on how fast the
compilation of the loadable builtins fails. Try the following to
reproduce the issue:

        nix build --rebuild --cores 48 nixpkgs#pkgsStatic.bash

This commit skips building and installing the loadable builtins
altogether if building statically.

alyssais commented:
Awesome, thanks! I'd like to wait a little bit to see what upstream says, in case we can avoid carrying the patch.
